File: apps/divsoup/lib/divsoup/analyzer/achievements/div_soup/bronze.ex
defmodule Divsoup.Achievement.DivSoupBronze do
  alias Divsoup.Achievement

  @behaviour Divsoup.AchievementRule

  @impl true
  def evaluate(html_tree, _) do
    ratio = Divsoup.Util.DivSoup.get_div_ratio(html_tree)

    if ratio > 0.25 do
      []
    else
      ["Only #{ratio * 100}% of the HTML elements in the page are divs"]
    end
  end

  @impl true
  def achievement() do
    %Achievement{
      hierarchy: :bronze,
      title: "Div Broth",
      group: "div_soup",
      description: "More than 25% of the HTML elements in the page are divs"
    }
  end
end

File: src/achievements/div_soup/bronze.ts
import type { AchievementRule } from "../types";
import { evaluateRule } from "../evaluate";

export const rule: AchievementRule = {
  id: "div_soup.bronze",
  title: "Div Broth",
  group: "div_soup",
  description:
    "More than <strong>25%</strong> of the HTML elements in the page are <code>&lt;div&gt;</code> elements",
  hierarchy: "bronze",
  evaluate: (context) => evaluateRule("div_soup.bronze", context),
};
